File: nodejs/extract-pdf-text-async.js
#!/usr/bin/env node

// TODO: You MUST install the following dependent module before running this script!
// npm install request
var request = require('request');


// TODO: When you have your own Client ID and secret, put down their values here:
var clientId = "FREE_TRIAL_ACCOUNT";
var clientSecret = "PUBLIC_SECRET";

// TODO: Specify the URL of your PDF document 
var url = "https://www.plainenglish.co.uk/files/partsofspeech.pdf";
var headers = {
      "X-WM-CLIENT-ID": clientId,
      "X-WM-CLIENT-SECRET": clientSecret,
    }

var optionsTemplate = {
  url: "TBD",
  headers: headers
};



function printError(body) {
    console.log("Something went wrong!!!");
    console.log(body);
}


function retrievePdfText(jobId) {
    var optionsRetrieve = JSON.parse(JSON.stringify(optionsTemplate));
    optionsRetrieve.url = 'https://api.whatsmate.net/v1/pdf/job/retrieve_text/' + jobId;

    request.get(optionsRetrieve, function (error, response, body) {
        if (response.statusCode == 200) {
            console.log("Here is the text extracted from PDF \n\n");
            console.log(body);
            // Idea: You can do something else with the extracted text. E.g. save the text in your database.
        } else {
            printError(body);
        }
        return;
    });
}


function checkJobStatus(jobId) {
    var optionsCheck = JSON.parse(JSON.stringify(optionsTemplate));
    optionsCheck.url = 'https://api.whatsmate.net/v1/pdf/job/check/' + jobId;
    request.get(optionsCheck, function (error, response, body) {
        if (response.statusCode == 200) {
            statusJson = JSON.parse(body)
            strStatus = statusJson.status

            if (strStatus == "Completed") {
                retrievePdfText(jobId);
                return;  // exit
            } else if (strStatus.startsWith("Failed")) {
                printError(body);
                return;  // exit
            }

            console.log("Need to wait a little more. Status: " + strStatus);
            setTimeout(checkJobStatus.bind(null, jobId), 60 * 1000);
        } else {
            printError(body);
            return;  // exit
        }

    });
}


function main() {
    var jobId = null;

    var optionsSubmitJob = JSON.parse(JSON.stringify(optionsTemplate));
    optionsSubmitJob.url = 'https://api.whatsmate.net/v1/pdf/job/submit?url=' + url;
    request.get(optionsSubmitJob, function (error, response, body) {
        if (response.statusCode == 200) {
            var jobJson = JSON.parse(body);
            jobId = jobJson.id;
            console.log('Job ID is : ' + jobId);

            console.log("Sleeping for 60 seconds");
            setTimeout(checkJobStatus.bind(null, jobId), 60 * 1000);
        } else {
            printError(body);
        }
    });
}


main();

File: python/extract-pdf-text-async.py
#!/usr/bin/env python
import time
import json
import requests

# TODO: When you have your own Client ID and secret, put down their values here:
clientId = "FREE_TRIAL_ACCOUNT"
clientSecret = "PUBLIC_SECRET"

# TODO: Specify the URL of your PDF document
url = "https://www.plainenglish.co.uk/files/partsofspeech.pdf"

headers = {
    'X-WM-CLIENT-ID': clientId, 
    'X-WM-CLIENT-SECRET': clientSecret
}

# Submit a long-running job for extracting text from PDF
r = requests.get('https://api.whatsmate.net/v1/pdf/job/submit?url=' + url, 
    headers=headers)

strJobJson = str(r.content)
print "Job JSON: " + strJobJson

jobJson = json.loads(strJobJson)
jobId = jobJson.get("id")
strStatus = jobJson.get("status")
strErr = jobJson.get("error_message")

if strErr is not None:
    print "Something is wrong: "
    print strErr
    quit()

# Check the job status regularly
doneProcessing = False
while not doneProcessing:
    print "Sleeping for 60 seconds"
    time.sleep(60)

    r = requests.get('https://api.whatsmate.net/v1/pdf/job/check/' + jobId, 
        headers=headers)
    statusJson = json.loads(str(r.content))
    strStatus = statusJson.get("status")

    if strStatus == "Completed":
        doneProcessing = True
    elif strStatus[:6] == "Failed":
        doneProcessing = True
        print "Something went wrong! Status was: " + strStatus
        quit()


# Retrieve the text
r = requests.get('https://api.whatsmate.net/v1/pdf/job/retrieve_text/' + jobId, 
    headers=headers)
print str(r.content)
